fv3_cap.F90

```diff
@@ -0,0 +1,1691 @@
+    if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) return
+
+    call NUOPC_CompSpecialize(gcomp, specLabel=model_label_CheckImport, &
+                              specPhaseLabel="phase2", specRoutine=NUOPC_NoOp, rc=rc)
+    if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) return
+
+    ! model finalize method(s)
+    call NUOPC_CompSpecialize(gcomp, specLabel=model_label_Finalize, &
+                              specRoutine=atmos_model_finalize, rc=rc)
+    if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) return
+
+  end subroutine SetServices
+
+  !-----------------------------------------------------------------------------
+
+  subroutine InitializeP0(gcomp, importState, exportState, clock, rc)
+    type(ESMF_GridComp)   :: gcomp
+    type(ESMF_State)      :: importState, exportState
+    type(ESMF_Clock)      :: clock
+    integer, intent(out)  :: rc
+
+    character(len=10)     :: value
+    character(240)        :: msgString
+    logical               :: isPresent, isSet
+    character(len=*),parameter  :: subname='(fv3gfs_cap:InitializeP0)'
+
+    rc = ESMF_SUCCESS
+
+    ! Switch to IPDv01 by filtering all other phaseMap entries
+    call NUOPC_CompFilterPhaseMap(gcomp, ESMF_METHOD_INITIALIZE, &
+                                  acceptStringList=(/"IPDv01p"/), rc=rc)
+    if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) return
+
+    call ESMF_AttributeGet(gcomp, name="ProfileMemory", value=value, defaultValue="true", &
+                           convention="NUOPC", purpose="Instance", rc=rc)
+    if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) return
+    profile_memory = (trim(value)/="false")
+
+    call ESMF_AttributeGet(gcomp, name="DumpFields", value=value, defaultValue="false", &
+                           convention="NUOPC", purpose="Instance", rc=rc)
+    if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) return
+
+    cplprint_flag = (trim(value)=="true")
+    write(msgString,'(A,l6)') trim(subname)//' cplprint_flag = ',cplprint_flag
+    call ESMF_LogWrite(trim(msgString), ESMF_LOGMSG_INFO, rc=rc)
+
+    ! Read in cap debug flag
+    call NUOPC_CompAttributeGet(gcomp, name='dbug_flag', value=value, isPresent=isPresent, isSet=isSet, rc=rc)
+    if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) return
+    if (isPresent .and. isSet) then
+     read(value,*) dbug
+    end if
+    write(msgString,'(A,i6)') trim(subname)//' dbug = ',dbug
+    call ESMF_LogWrite(trim(msgString), ESMF_LOGMSG_INFO, rc=rc)
+
+  end subroutine
+
+  !-----------------------------------------------------------------------------
+
+  subroutine InitializeAdvertise(gcomp, importState, exportState, clock, rc)
```

Comment:
This is a very long subroutine that is doing a lot of stuff.  It should be broken into several subroutines, each with a clear purpose.

Lines 289-483:  Reading in configuration.  Much of the config has to do with output (e.g., kind of output grid) and probably should go in the output component itself.

Lines 509-562:  Reading and setting up restart, including changes to the clock.

Lines 573-795:  Creating forecast and write component and regridding routehandles

Lines 801-870:  Creating output alarms

The last few lines do the actual advertise of fields.

Reply:
Is this a standard code style? We got different feedbacks on this.